### PR TITLE
SYNR-1275 SYNR-1322 SYNR-1325

### PR DESCRIPTION
**Pull request by nickgros-sage**
Coerce to and from specific R types for tables instead of letting read.csv guess

### CODE REVIEW

File: R/table.R [hunk 1]
```diff
@@ -12,11 +12,11 @@
     for (i in 1:dim(dataFrame)[2]) {
       if (is.numeric(dataFrame[[i]])) {
         dataFrame[[i]][is.nan(dataFrame[[i]])] <- "NaN"
-      } else if (is(dataFrame[[i]], "POSIXct")) {
-        # convert POSIXct before uploading to Synapse
+      } else if (is(dataFrame[[i]], "POSIXt")) {
+        # convert POSIX time to the same precision offered by Synapse
         dataFrame[[i]] <- format(
           as.POSIXlt(dataFrame[[i]], "UTC", usetz = TRUE),
```
Comment (nickgros-sage) [inline R/table.R]: Convert to timestamp

File: R/table.R [hunk 2]
```diff
@@ -42,3 +43,108 @@
...
+    } else if (is(list, "numeric")) {
+      list
+    } else {
+      stop(paste("Cannot convert type ", class(list), "to a ", synapseType, "."))
+    }
+  } else if (synapseType == "INTEGER"){
+    tryCatch(
+      as.integer(list),
+      warning = function(x) { as.numeric(list) } # in case the integers are outside of the bounds of R integer
+    )
```
Comment (nickgros-sage) [inline R/table.R]: check format in test (trailing 0?)

File: R/table.R [hunk 3]
```diff
@@ -42,3 +43,108 @@
     }
   )
 }
+
```
Comment (nickgros-sage) [inline R/table.R]: Add to vignette
- modify example by adding Date
- mention handling date in a new section (reference POSIXlt docs)

File: tests/testthat/test_table.R [hunk 4]
```diff
@@ -89,3 +89,474 @@ test_that("as.data.frame works for CsvFileTable", {
...
+  expected <- c("273954", "273950", NA)
+  
+  actual <- .convertToRType(list, type)
+  
+  expect_is(actual, "character")
+  expect_equal(expected, actual)
+})
+
+test_that(".convertToRType works for DOUBLE", {
+  list <- c("3", "900", "")
```
Comment (nickgros-sage) [inline tests/testthat/test_table.R]: all cases add NA, NULL

File: R/table.R [hunk 5]
```diff
@@ -42,3 +43,108 @@
...
+# Converts data downloaded from Synapse to an appropriate data type in R
+.convertToRType <- function(list, synapseType) {
```
Comment (nickgros-sage) [inline R/table.R]: create test(s) for converting back and forth and assuring the converted type is the same as the original

File: tests/testthat/test_table.R [hunk 6]
```diff
@@ -89,3 +89,474 @@ test_that("as.data.frame works for CsvFileTable", {
...
+  
+  actual <- .convertToRType(list, type)
+  
+  expect_is(actual, "numeric")
+  expect_false(is(actual, "integer"))
+  expect_equal(expected, actual)
+})
+
+test_that(".convertToRType works for STRING", {
+  list <- c("42", "24.24", NA, "NULL", "NA", "")
```
[the remark this answers is not in the document]
Comment (nickgros-sage) [reply tests/testthat/test_table.R]: covered in another test

File: R/table.R [hunk 7]
```diff
@@ -13,11 +13,10 @@
       if (is.numeric(dataFrame[[i]])) {
         dataFrame[[i]][is.nan(dataFrame[[i]])] <- "NaN"
       } else if (is(dataFrame[[i]], "POSIXt")) {
-        # convert POSIX time to the same precision offered by Synapse
-        dataFrame[[i]] <- format(
-          as.POSIXlt(dataFrame[[i]], "UTC", usetz = TRUE),
-          "%Y-%m-%d %H:%M:%OS3"
-        )
+        # convert POSIX time to timestamp
+        dataFrame[[i]] <- trimws(format(as.numeric(dataFrame[[i]]) * 1000, scientific = FALSE))
```
Comment (kimyen) [inline R/table.R]: extract line 17 - 19 to a function

File: vignettes/tables.Rmd [hunk 8]
```diff
@@ -198,6 +202,33 @@ To access a column that you do not have a reference to, please see:
...
+
+Dates may also be submitted in timestamp milliseconds:
+```{r collapse=T}
+results <- synTableQuery(sprintf("select * from %s where Name='zap'", tableId))
+df <- as.data.frame(results)
+df$Time <- timestamp
+synStore(Table(tableId, moreGenes))
+```
+
+```
```
Comment (kimyen) [inline vignettes/tables.Rmd]: remove this

**Pull request merged**